combat/CombatDamage.cpp

```diff
@@ -0,0 +1,211 @@
+        //auto target = std::make_shared<Ship>(ALL_EMPIRES, GetPredefinedShipDesignManager().GetDesignID("SD_MARK_1"), "SP_HUMAN");
+        //auto target = std::make_shared<Fighter>();
+        // XXX DesignID valid?
+        auto target = std::make_shared<Ship>(ALL_EMPIRES, source->DesignID(), "SP_HUMAN");
+        // target needs to have an ID != -1 to be visible, stealth should be low enough
+        // structure must be higher than zero to be valid target
+        target->SetID(-1000000); // XXX magic number AutoresolveInfo.next_fighter_id starts at -1000001 counting down
+        target->GetMeter(MeterType::METER_STRUCTURE)->Set(100.0f, 100.0f);
+        target->GetMeter(MeterType::METER_MAX_STRUCTURE)->Set(100.0f, 100.0f);
+        // Shield value is used for structural damage estimation
+        target->GetMeter(MeterType::METER_SHIELD)->Set(shields,shields);
+        GetUniverse().SetEmpireObjectVisibility(source->Owner(), target->ID(), Visibility::VIS_FULL_VISIBILITY);
+        return ScriptingContext(source, target);
+    }
+
+    ScriptingContext CombatContextWithFighterTarget(std::shared_ptr<const Ship> source) {
+        auto target = std::make_shared<Fighter>();
+        target->SetID(-1000000); // XXX magic number AutoresolveInfo.next_fighter_id starts at -1000001 counting down
+        GetUniverse().SetEmpireObjectVisibility(source->Owner(), target->ID(), Visibility::VIS_FULL_VISIBILITY);
+        return ScriptingContext(source, target);
+    }
+
+    std::vector<float> WeaponDamageCalcImpl(std::shared_ptr<const Ship> ship, const ShipDesign* design,
+                                            bool max, bool include_fighters, bool target_ships, const ScriptingContext&& context) {
+        std::vector<float> retval;
+        if (!ship || !design)
+            return retval;
+        const std::vector<std::string>& parts = design->Parts();
+        if (parts.empty())
+            return retval;
+
+        MeterType METER = max ? MeterType::METER_MAX_CAPACITY : MeterType::METER_CAPACITY;
+        MeterType SECONDARY_METER = max ? MeterType::METER_MAX_SECONDARY_STAT : MeterType::METER_SECONDARY_STAT;
+
+        float fighter_damage = 0.0f;
+        int fighter_launch_capacity = 0;
+        int available_fighters = 0;
+
+        retval.reserve(parts.size() + 1);
+        int num_bouts = GetGameRules().Get<int>("RULE_NUM_COMBAT_ROUNDS");
+        // for each weapon part, get its damage meter value
+        for (const auto& part_name : parts) {
+            const ShipPart* part = GetShipPart(part_name);
+            if (!part)
+                continue;
+            ShipPartClass part_class = part->Class();
+
+            // get the attack power for each weapon part.
+            if (part_class == ShipPartClass::PC_DIRECT_WEAPON) {
+                if (target_ships)
+                    retval.emplace_back(ship->WeaponPartShipDamage(part, context));
```

Comment:
prefer `push_back` to `emplace_back` for a single existing object. use `emplace_back` to forward parameters to a constructor